File: 1466-jump-game-v/jump-game-v.java
import java.util.*;

class Solution {

    public int maxJumps(int[] arr, int d) {

        int n = arr.length;
        int[] dp = new int[n];

        // har index se minimum 1 jump
        for (int i = 0; i < n; i++) {
            dp[i] = 1;
        }

        // index array
        int[] idx = new int[n];
        for (int i = 0; i < n; i++) {
            idx[i] = i;
        }

        // 🔁 simple sorting (bubble sort) based on height
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[idx[j]] > arr[idx[j + 1]]) {
                    int temp = idx[j];
                    idx[j] = idx[j + 1];
                    idx[j + 1] = temp;
                }
            }
        }

        // bottom-up DP
        for (int x = 0; x < n; x++) {

            int id = idx[x];

            // left side
            for (int j = id - 1; j >= 0 && j >= id - d; j--) {
                if (arr[j] >= arr[id])
                    break;
                dp[id] = Math.max(dp[id], 1 + dp[j]);
            }

            // right side
            for (int j = id + 1; j < n && j <= id + d; j++) {
                if (arr[j] >= arr[id])
                    break;
                dp[id] = Math.max(dp[id], 1 + dp[j]);
            }
        }

        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, dp[i]);
        }

        return ans;
    }
}

File: 1466-jump-game-v/jump-game-v.py
class Solution(object):
    def maxJumps(self, arr, d):
        """
        :type arr: List[int]
        :type d: int
        :rtype: int
        """
        n = len(arr)
        dp = [0] * n  # memoization (0 means uncomputed)

        def dfs(i):
            if dp[i] != 0:
                return dp[i]

            res = 1  # Minimum (stay in place)

            # Move left
            for j in range(i - 1, i - d - 1, -1):
                if j < 0 or arr[j] >= arr[i]:
                    break
                res = max(res, 1 + dfs(j))

            # Move right
            for j in range(i + 1, i + d + 1):
                if j >= n or arr[j] >= arr[i]:
                    break
                res = max(res, 1 + dfs(j))

            dp[i] = res
            return res

        ans = 0
        for i in range(n):
            ans = max(ans, dfs(i))

        return ans


    